### PR TITLE
release,cmd,dirs: Redo the distro checks to take into account distribution families

### PR DESCRIPTION
**Pull request by Conan-Kudo**
This PR changes the distribution checks in snapd to account for distribution families, so that snapd behaves as expected across not only the main distributions we support, but their derivatives, too.

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -66,8 +66,17 @@ func distroSupportsReExec() bool {
...
+	var contains = func(haystack []string, needle string) bool {
+		for _, s := range haystack {
+			if s == needle {
+				return true
+			}
+		}
+		return false
+
+	}
+	switch {
```
Comment (zyga) [inline cmd/cmd.go]: The switch makes little sense here, can you make that into a normal if statement now please?

File: cmd/cmd.go [hunk 2]
```diff
@@ -66,8 +66,17 @@ func distroSupportsReExec() bool {
 	if !release.OnClassic {
 		return false
 	}
-	switch release.ReleaseInfo.ID {
-	case "fedora", "centos", "rhel", "opensuse", "suse", "poky", "arch", "manjaro", "lirios":
+	var contains = func(haystack []string, needle string) bool {
```
Comment (zyga) [inline cmd/cmd.go]: Can you please move this to `strutil.Contains`?

File: dirs/dirs.go [hunk 3]
```diff
@@ -165,8 +165,17 @@ func SetRootDir(rootdir string) {
 	}
 	GlobalRootDir = rootdir
 
-	switch release.ReleaseInfo.ID {
-	case "fedora", "centos", "rhel", "arch", "manjaro", "lirios":
+	var contains = func(haystack []string, needle string) bool {
```
Comment (zyga) [inline dirs/dirs.go]: Once you move the other function please just reuse it here.

File: release/release_test.go [hunk 4]
```diff
@@ -93,6 +93,35 @@ BUG_REPORT_URL="https://bugs.launchpad.net/elementary/+filebug"`
 	c.Check(os.VersionID, Equals, "0.4")
 }
 
+func (s *ReleaseTestSuite) TestFamilyOSRelease(c *C) {
+	mockOSRelease := filepath.Join(c.MkDir(), "mock-os-release")
```
Comment (zyga) [inline release/release_test.go]: Nice test case

File: release/release_test.go [hunk 5]
```diff
@@ -93,6 +93,35 @@ BUG_REPORT_URL="https://bugs.launchpad.net/elementary/+filebug"`
...
+	dump := `NAME="CentOS Linux"
+	VERSION="7 (Core)"
```
Comment (zyga) [inline release/release_test.go]: I think you have to unindent this. Alternatively just move the variable dump out of the function (and make it a const)

File: cmd/cmd.go (hunk 2 above)
Comment (Conan-Kudo) [reply cmd/cmd.go]: Hmm, apparently this already exist as `strutil.ListContains`, I'll just reuse that.

File: cmd/cmd.go (hunk 1 above)
Comment (Conan-Kudo) [reply cmd/cmd.go]: This is still checking more than one conditional, so the switch makes sense still, I think.

**Comment by Conan-Kudo** [comment]
@zyga This should address all your feedback now. 👍

File: cmd/cmd.go (hunk 2 above)
Comment (zyga) [reply cmd/cmd.go]: Ah, perfect.

**Comment by Conan-Kudo** [comment]
I've rebased it against current master to resolve conflicts.

File: cmd/cmd.go [hunk 6]
```diff
@@ -66,8 +66,8 @@ func distroSupportsReExec() bool {
 	if !release.OnClassic {
 		return false
 	}
-	switch release.ReleaseInfo.ID {
-	case "fedora", "centos", "rhel", "opensuse", "suse", "poky", "arch", "manjaro", "lirios", "solus":
+	switch {
+	case release.ReleaseInfo.ID != "debian", !strutil.ListContains(release.ReleaseInfo.IDLike, "debian"):
```
Comment (zyga) [inline cmd/cmd.go]: Can you please add a similar check for `ID_LIKE=ubuntu`, the code as it stands today would regress on Elementary OS
```
NAME="elementary OS"
VERSION="0.4 Loki"
ID="elementary OS"
ID_LIKE=ubuntu
PRETTY_NAME="elementary OS Loki"
VERSION_ID="0.4"
HOME_URL="http://elementary.io/"
SUPPORT_URL="http://elementary.io/support/"
BUG_REPORT_URL="https://bugs.launchpad.net/elementary/+filebug"
```

File: release/release_test.go [hunk 7]
```diff
@@ -93,6 +93,35 @@ BUG_REPORT_URL="https://bugs.launchpad.net/elementary/+filebug"`
...
+
+	reset := release.MockOSReleasePath(mockOSRelease)
+	defer reset()
+
+	os := release.ReadOSRelease()
+	c.Check(os.ID, Equals, "centos")
+	c.Check(os.VersionID, Equals, "7")
+	c.Check(os.IDLike, DeepEquals, []string{"rhel", "fedora"})
+}
+
```
Comment (zyga) [inline release/release_test.go]: I'd like to see a few more test cases for distributions:
- fedora
- centos (thanks)
- rhel
- ubuntu
- ubuntu core
- elementary
- mint
- arch

etc.

What we save by having nice generic code here I'd like to check by having very comprehensive test suite that measures behaviour.

File: cmd/cmd.go (hunk 6 above)
Comment (mvo5) [inline cmd/cmd.go]: It seems like an "if" statement here should be fine, a switch {} with just a single case looks a bit out of place.

File: release/release_test.go (hunk 7 above)
Comment (mvo5) [reply release/release_test.go]: AIUI @zyga has a hole bunch of examples of those maybe we can import them here?

Comment (zyga) [reply release/release_test.go]: Yes, I linked to them above.

Comment (Conan-Kudo) [reply release/release_test.go]: I don't want to import all of them, as that isn't quite necessary. We just need to prove that the negative cases work.

File: dirs/dirs.go [hunk 8]
```diff
@@ -166,8 +167,8 @@ func SetRootDir(rootdir string) {
 	}
 	GlobalRootDir = rootdir
 
-	switch release.ReleaseInfo.ID {
-	case "fedora", "centos", "rhel", "arch", "manjaro", "lirios":
+	switch {
+	case release.ReleaseInfo.ID == "fedora", strutil.ListContains(release.ReleaseInfo.IDLike, "fedora"), release.ReleaseInfo.ID == "arch", strutil.ListContains(release.ReleaseInfo.IDLike, "arch"):
```
Comment (zyga) [inline dirs/dirs.go]: I'm not sure if Manjaro uses `/var/lib/snapd/snap` anymore, I think it was reverted back to `/snap` there

File: cmd/cmd.go [hunk 9]
```diff
@@ -66,8 +66,8 @@ func distroSupportsReExec() bool {
 	if !release.OnClassic {
 		return false
 	}
-	switch release.ReleaseInfo.ID {
-	case "fedora", "centos", "rhel", "opensuse", "suse", "poky", "arch", "manjaro", "lirios", "solus":
+	if release.ReleaseInfo.ID != "debian" || !strutil.ListContains(release.ReleaseInfo.IDLike, "debian") ||
```
Comment (zyga) [inline cmd/cmd.go]: Thank you

**Comment by Conan-Kudo** [comment]
@zyga @mvo5 Help with the tests would be appreciated! I'm at my wit's end with trying to make this pass, and the Wi-Fi is behaving weird now...

File: dirs/dirs.go (hunk 8 above)
Comment (Conan-Kudo) [reply dirs/dirs.go]: Manjaro doesn't identify as like Arch, so that should be fine.

**Comment by Conan-Kudo** [comment]
@zyga @mvo5 This is now passing all tests, so can it be merged?

Also, please cherry pick this PR into 2.28, because it'll fix some big issues with Fedora derivatives.

**Pull request merged**